library/core/src/future/poll_fn.rs

```diff
@@ -24,6 +24,75 @@ use crate::task::{Context, Poll};
 /// assert_eq!(read_future.await, "Hello, World!".to_owned());
 /// # }
 /// ```
+///
+/// ## Capturing a pinned state
+///
+/// Example of a closure wrapping an inner future:
+///
+/// ```
+/// # async fn run() {
+/// use core::future::{Future, poll_fn};
+/// use core::mem;
+/// use core::task::Poll;
+///
+/// /// Wraps a `Future` so that it yields `Pending` one time before forwarding to the inner
+/// /// `Future`.
+/// fn yield_once_then<T>(fut: impl Future<Output = T>) -> impl Future<Output = T> {
+///     let mut should_yield = true;
+///     let mut pinned_fut = Box::pin(fut);
+///     poll_fn(move |cx| if mem::take(&mut should_yield) {
+///         // Remember: for a `Future` impl to be correct, any `Pending` return has to have,
+///         // somewhere, a matching `wake` call.
```

Comment:
This comment is a tad confusing. I think it'd read a bit more cleanly if we said something like:
> Immediately reschedule this Future for execution by calling `wake` before returning the initial `Pending`.

Given the phrasing and the structure of the code, my impression was that `wake` _must_ be called before returning `Pending` which isn't strictly correct in all circumstances. `wake` should only be called when we want the Future to be scheduled for execution.

Reply:
I guess this is what the _somewhere_ was trying to convey, but I agree it's not 100% clear. I'd still like to keep the _mandatory_ aspect of my phrasing (since I've seen many people mis-implement `Future`s when returning `Pending` manually).

What about:

```suggestion
///         // Remember: for a `Future` impl to be correct, any `Pending` return has to have,
///         // once ready, a matching `wake` call, lest it not be rescheduled for execution.
```

Reply:
I think this is a lot better, yeah. We should consider this resolved.